### PR TITLE
SidebarList component

### PR DESCRIPTION
**Pull request by swese44**
https://app.zeplin.io/project/5907d6451d2b06c6b36b2852/screen/59c582c900581ab06b104f2f
Drag n drop sorting and remove button are not implemented. We can add this functionality when needed.

## Pull request checklist

* [x] Component `README.md` file is up-to-date.
* [x] Component is unit tested.

### CODE REVIEW

File: src/components/SidebarList/SidebarListItem.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import { getClassNames, SidebarListItemClassNames } from './SidebarListItem.styles';
+
+export { SidebarListItem };
+
+const getIllustrationSize = (size: SidebarListItemProps['size']) => (size === 'large' ? 32 : 24);
+const getMediaObjectSize = (size: SidebarListItemProps['size']) =>
+  size === 'large' ? MediaObjectSize.SMALL : MediaObjectSize.XSMALL;
+
+const isLinkItem = (item: SidebarListItem): item is SidebarListLinkItem => {
+  return (item as SidebarListLinkItem).href !== undefined;
```
[the remark this answers is not in the document]
Comment (swese44) [reply src/components/SidebarList/SidebarListItem.tsx]: i think this is the tradeoff to allow a simpler API surface to consumers while still providing complex union type safety.

this instance is actually a best practice of defining a type guard with the `is` keyword, and lets the rest of the file call `isLinkItem` without typecasting every time to check props:
http://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

i can add another typeguard for `isIllustrationItem` so `getImageContent` doesn't need to typecast.

the last cast in this file is in `onClick`. i don't think we can avoid it while maintaining 100% code coverage, but it's safe and probably a good example of when it's ok to type case

File: src/components/SidebarList/SidebarList.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  public render() {
+    return (
+      <LayoutList direction="vertical" className={join(['y-sidebarList', this.props.className])}>
+        {this.getItems()}
+      </LayoutList>
+    );
+  }
+
+  private getItems() {
+    return this.props.items.map(item => <SidebarListItemView item={item} size={this.props.size} key={item.key} />);
```
[the remark this answers is not in the document]
Comment (richardtagger) [reply src/components/SidebarList/SidebarList.tsx]: avoid the function being recreated ever re-render. Dont feel super strong but its something we do quite a lot

[the remark this answers is not in the document]
Comment (swese44) [reply src/components/SidebarList/SidebarList.tsx]: ah i follow. good call, i'll update

[the remark this answers is not in the document]
Comment (swese44) [reply src/components/SidebarList/SidebarList.tsx]: fixed

File: src/components/SidebarList/SidebarListItem.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (swese44) [reply src/components/SidebarList/SidebarListItem.tsx]: updated. only `as` typecasting is in the 2 `is` type guard functions, and the onClick handler

**Comment by swese44** [comment]
oh just saw your comment on keyboard nav. at the moment things are fine, tabbing focuses each wrapping button or link item. if we implement sorting or close we'll need a little more work.

**Pull request merged**